---
title: "systemPipeR"
author: "Author: Thomas Girke (thomas.girke@ucr.edu)"
date: "July 22, 2015" 
output:
  ioslides_presentation:
    keep_md: yes
    logo: logo.png
    toc: yes
    toc_depth: 2
    widescreen: yes
subtitle: "NGS workflow and report generation environment" 
bibliography: bibtex.bib
---

<!---
- Latex to markdown conversion with pandoc
  pandoc -s systemPipeRslides.tex -o systemPipeRslides.text

- rmarkdown manuals/movies: 
  http://rmarkdown.rstudio.com/ioslides_presentation_format.html
	http://blog.rstudio.org/2014/06/18/r-markdown-v2/

- Compile from R console like this
library(rmarkdown)
render('systemPipeRslides.Rmd', ioslides_presentation(logo="logo.png"))

- Compile from command-line
echo "library(rmarkdown); library(knitr); render('systemPipeRslides.Rmd', ioslides_presentation(logo='logo.png')); knit('systemPipeRslides.Rmd', tangle=TRUE)" | R -slave
-->

Outline
=======================
- <div class="white">__Introduction__</div>
- Motivation
- Design
- Getting started

## Introduction 

-   [_systemPipeR_](http://www.bioconductor.org/packages/devel/bioc/html/systemPipeR.html) is an R package for building *end-to-end* analysis pipelines with automated report generation for next generation NGS applications [@Girke2014-oy].
-   Important features:
    - Support for R and command-line software, such as NGS aligners, peak callers, variant callers, etc.
    - Runs on single machines and compute clusters with schedulers
    - Uniform sample handling and annotation

Outline
=======================
- Introduction
- <div class="white">__Motivation__</div>
- Design
- Getting started

## Motivation and advantages 

- Facilitates design of complex NGS workflows involving multiple R/Bioconductor packages [@Huber2015-ag].
- Makes NGS analysis with Bioconductor utilities more accessible to new users
- Simplifies usage of command-line software from within R
- Reduces complexity of using compute clusters for R and command-line software
- Accelerates runtime of workflows via parallelzation on computer systems with mutiple CPU cores and/or multiple compute nodes
- Automates generation of analysis reports to improve reproducibility

Outline
=======================
- Introduction
- Motivation
- <div class="white">__Design__</div>
- Getting started

## Workflow design in _`systemPipeR`_ {.flexbox .vcenter .smaller}

<img src="./images/SystemPipeR_Workflow.png" alt="Drawing" style="width: 800px;"/>
</div><br></br>

- Workflow steps with input/output file operations are controlled by _`SYSargs`_ objects.

- Each _`SYSargs`_ instance is constructed from a _`targets`_ file and a _`param`_ file.

- Only input provided by user is initial _`targets`_ file. Subsequent _`targets`_ instances are created automatically. 

- Any number of predefined or custom workflow steps are supported.


Outline
=======================
- Introduction
- Motivation
- Design
- <div class="white">__Getting Started__</div>

## Install and load packages {.smaller}

Install required packages
```{r install, eval=FALSE}
source("http://bioconductor.org/biocLite.R") 
biocLite("systemPipeR")
biocLite("tgirke/systemPipeRdata", build_vignettes=TRUE, dependencies=TRUE) # From github
```

Load packages
```{r load_systemPipeR_hidden, eval=TRUE, include=FALSE, messages=FALSE, warnings=FALSE, cache=TRUE}
library("systemPipeR"); library("systemPipeRdata")
```
```{r load_systemPipeR_print, eval=FALSE, messages=FALSE, warnings=FALSE, cache=TRUE}
library("systemPipeR"); library("systemPipeRdata")
```

Structure of _`targets`_ file
```{r show_targets, eval=TRUE, messages=FALSE, warnings=FALSE, cache=TRUE}
targetspath <- system.file("extdata", "targets.txt", package="systemPipeR")
read.delim(targetspath, comment.char = "#")[1:4,1:5]
```

## Workflow templates 

Generate workflow template, _e.g._ "rnaseq", "varseq" or "chipseq"
```{r generate_workenvir, eval=FALSE, cache=TRUE}
### <b>
genWorkenvir(workflow="varseq", mydirname=NULL)
### </b>
setwd("varseq")
```
<br></br>
Command-line alternative for generating workflow environments
```{.sh generate_workenvir_from_shell, eval=FALSE, cache=TRUE}
$ echo 'library(systemPipeRdata); 
        genWorkenvir(workflow="varseq", mydirname=NULL)' | R --slave
```

## Workflow template structure

The workflow templates generated by _`genWorkenvir`_ contain the following preconfigured directory structure:
<br></br>
```{r workflow_template_structure, eval=FALSE}
workflow_name/            # *.Rnw/*.Rmd scripts and targets file
                param/    # parameter files for command-line software 
                data/     # inputs e.g. FASTQ, reference, annotations
                results/  # analysis result files
```
<br></br>
The above structure can be customized as needed, but for first time users it is easier to leave it as is.

## Run workflows 

- Next, run from within R the chosen sample workflow by executing the code provided in the corresponding _`*.Rnw`_ template file (or _`*.Rmd`_ or _`*.R`_ versions).
- Alternatively, one can run an entire workflow from start to finish with a single command by executing from the command-line:
```{.sh run_make, eval=FALSE}
$ make -B
```

## Continue here  
<br></br><br></br><br></br>
<center>
[Overview Vignette](http://www.bioconductor.org/packages/devel/bioc/vignettes/systemPipeR/inst/doc/systemPipeR.html) 
</center>

## References {.smaller}


